lib/sles4sap/sdaf_library.pm

```diff
@@ -0,0 +1,519 @@
+
+SPN credentials are defined by secret OpenQA parameters:
+
+B<_SECRET_AZURE_SDAF_APP_ID>
+
+B<_SECRET_AZURE_SDAF_APP_PASSWORD>
+
+B<_SECRET_AZURE_SDAF_TENANT_ID>
+
+SDAF needs SPN credentials with special permissions. Check link below for details.
+https://learn.microsoft.com/en-us/azure/sap/automation/deploy-control-plane?tabs=linux#prepare-the-deployment-credentials
+
+=cut
+
+sub az_login {
+    # SDAF tests execute on same jump VM, therefore each file needs unique ID
+    my $temp_file = '/tmp/az_login_' . get_current_job_id();
+    my @variables = (
+        'export ARM_CLIENT_ID=' . get_required_var('_SECRET_AZURE_SDAF_APP_ID'),
+        'export ARM_CLIENT_SECRET=' . get_required_var('_SECRET_AZURE_SDAF_APP_PASSWORD'),
+        'export ARM_TENANT_ID=' . get_required_var('_SECRET_AZURE_SDAF_TENANT_ID'),
+    );
+
+    # Write variables into temporary file using openQA infrastructure to avoid exposing variable values.
+    write_sut_file($temp_file, join("\n", @variables));
+    # Source file and load variables
+    assert_script_run("source $temp_file");
+
+    my $login_cmd = 'while ! az login --service-principal -u ${ARM_CLIENT_ID} -p ${ARM_CLIENT_SECRET} -t ${ARM_TENANT_ID}; do sleep 10; done';
+    assert_script_run($login_cmd, timeout => 5 * 60);
+
+    my $subscription_id = script_output('az account show -o tsv --query id');
+    record_info('AZ login', "Subscription id: $subscription_id");
+
+    # Remove temp file with credentials.
+    assert_script_run("rm $temp_file");
+    return ($subscription_id);
+}
+
+=head2 create_sdaf_os_var_file
+
+    create_sdaf_os_var_file($entries);
+
+B<$entries>: ARRAYREF of entries to be appended to variable source file
+
+Creates a simple file with bash env variables and uploads it to the target host without revealing content in serial console.
+File is sourced afterwards.
+
+=cut
+
+sub create_sdaf_os_var_file {
+    my ($entries) = @_;
+    my @entry_list = @$entries;
+    write_sut_file(variable_file, join("\n", "\n", @entry_list));
+    assert_script_run('source ' . variable_file, quiet => 1);
+}
+
+=head2 sdaf_get_deployer_ip
+
+    sdaf_get_deployer_ip(deployer_resource_group=>$deployer_resource_group);
+
+B<deployer_resource_group>: Deployer key vault name
+
+Retrieves public IP of the deployer VM.
+
+=cut
+
+sub sdaf_get_deployer_ip {
+    my (%args) = @_;
+    croak 'Missing "deployer_resource_group" argument' unless $args{deployer_resource_group};
+
+    my $vm_name = script_output("az vm list --resource-group $args{deployer_resource_group} --query [].name --output tsv");
+    my $az_query_cmd = join(' ', 'az', 'vm', 'list-ip-addresses', '--resource-group', $args{deployer_resource_group},
+        '--name', $vm_name, '--query', '"[].virtualMachine.network.publicIpAddresses[0].ipAddress"', '-o', 'tsv');
+
+    my $ip_addr = script_output($az_query_cmd);
+    croak "Not a valid ip addr: $ip_addr" unless grep /^$RE{net}{IPv4}$/, $ip_addr;
```

Comment:
is it really needed? What kind of scenario this validation is for?

Reply:
It verifies if script_output returned IP addr or something additional. 
Technically I could leave test fail at later point while attempting to ssh but this will save the time.